File: src/agent/tools/builtin/shell.ts
import type { ToolHandler } from "../../../types";

export const shellTool: ToolHandler = {
  definition: {
    name: "run_shell",
    description:
      "Execute a shell command on the host machine. Returns stdout, stderr, and exit code. " +
      "Default working directory is the project root (where openagent.json lives). " +
      "Timeout is 10 seconds by default — keep commands short. " +
      "WARNING: Do NOT use 'find' without '-maxdepth' — it will scan the entire disk and hang. " +
      "Do NOT start/restart the gateway server (bun run start/daemon/dev) — " +
      "those are long-running processes. Use write_config for config changes.",
    parameters: {
      type: "object",
      properties: {
        command: {
          type: "string",
          description: "The shell command to execute",
        },
        cwd: {
          type: "string",
          description:
            "Working directory. Defaults to the project root directory.",
        },
        timeout: {
          type: "number",
          description: "Timeout in milliseconds. Defaults to 10000.",
        },
      },
      required: ["command"],
    },
  },

  async execute(args) {
    const command = args.command as string;
    const cwd = (args.cwd as string) || process.cwd();
    const timeout = (args.timeout as number) || 10_000;

    try {
      const proc = Bun.spawn(["sh", "-c", command], {
        cwd,
        stdout: "pipe",
        stderr: "pipe",
      });

      const timeoutId = setTimeout(() => proc.kill(), timeout);

      const [stdout, stderr] = await Promise.all([
        new Response(proc.stdout).text(),
        new Response(proc.stderr).text(),
      ]);

      clearTimeout(timeoutId);
      const exitCode = await proc.exited;

      return JSON.stringify({
        exitCode,
        stdout: stdout.slice(0, 10_000),
        stderr: stderr.slice(0, 5_000),
      });
    } catch (err) {
      return JSON.stringify({
        error: err instanceof Error ? err.message : String(err),
      });
    }
  },
};

File: src/tools/builtins/shell.py
"""Shell command execution tool, matching TS tools/builtin/shell.ts."""

from __future__ import annotations

import asyncio
import json
import os

from src.types import ToolDefinition, ToolHandler

PROJECT_ROOT = os.getcwd()
MAX_STDOUT = 10_240
MAX_STDERR = 5_120


async def _execute(args: dict) -> str:
    command = args.get("command", "")
    cwd = args.get("cwd", PROJECT_ROOT)
    timeout = args.get("timeout", 10)

    try:
        proc = await asyncio.create_subprocess_shell(
            command,
            cwd=cwd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        try:
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=timeout)
        except asyncio.TimeoutError:
            proc.kill()
            await proc.wait()
            return json.dumps({"error": "timeout", "timeout": timeout})

        return json.dumps({
            "exitCode": proc.returncode,
            "stdout": (stdout or b"").decode(errors="replace")[:MAX_STDOUT],
            "stderr": (stderr or b"").decode(errors="replace")[:MAX_STDERR],
        })
    except Exception as e:
        return json.dumps({"error": str(e)})


shell_tool = ToolHandler(
    definition=ToolDefinition(
        name="run_shell",
        description=(
            "Execute a shell command and return stdout/stderr. "
            "Use for system commands, file inspection, builds, etc. "
            "Warnings: do not use `find` without `-maxdepth`; do not start/restart servers."
        ),
        parameters={
            "type": "object",
            "properties": {
                "command": {
                    "type": "string",
                    "description": "The shell command to execute.",
                },
                "cwd": {
                    "type": "string",
                    "description": f"Working directory (defaults to project root: {PROJECT_ROOT}).",
                },
                "timeout": {
                    "type": "number",
                    "description": "Timeout in seconds (default: 10).",
                },
            },
            "required": ["command"],
        },
    ),
    execute=_execute,
)
